# Exercise

Use this air bnb new york city [data set](https://www.kaggle.com/dgomonov/new-york-city-airbnb-open-data/data) and remove outliers using percentile based on price per night for a given apartment/home. You can use suitable upper and lower limits on percentile based on your intuition. 
Your goal is to come up with new pandas dataframe that doesn't have the outliers present in it.

In [1]:
import pandas as pd

In [3]:
df = pd.read_csv("AB_NYC_2019.csv")

df.head()

id                                              name  host_id  \
0  2539                Clean & quiet apt home by the park     2787   
1  2595                             Skylit Midtown Castle     2845   
2  3647               THE VILLAGE OF HARLEM....NEW YORK !     4632   
3  3831                   Cozy Entire Floor of Brownstone     4869   
4  5022  Entire Apt: Spacious Studio/Loft by central park     7192   

     host_name neighbourhood_group neighbourhood  latitude  longitude  \
0         John            Brooklyn    Kensington  40.64749  -73.97237   
1     Jennifer           Manhattan       Midtown  40.75362  -73.98377   
2    Elisabeth           Manhattan        Harlem  40.80902  -73.94190   
3  LisaRoxanne            Brooklyn  Clinton Hill  40.68514  -73.95976   
4        Laura           Manhattan   East Harlem  40.79851  -73.94399   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0     Private room    149               1                  9  2018-10-19   
1  Entire home/apt    225               1                 45  2019-05-21   
2     Private room    150               3                  0         NaN   
3  Entire home/apt     89               1                270  2019-07-05   
4  Entire home/apt     80              10                  9  2018-11-19   

   reviews_per_month  calculated_host_listings_count  availability_365  
0               0.21                               6               365  
1               0.38                               2               355  
2                NaN                               1               365  
3               4.64                               1               194  
4               0.10                               1                 0

In [8]:
df.shape

(48895, 16)

48895 total entries

In [10]:
max_thresold = df['price'].quantile(0.99)
max_thresold

799.0

In [11]:
min_thresold = df['price'].quantile(0.01)
min_thresold

30.0

In [12]:
# these are outliers

df[(df['price']>=max_thresold) | (df['price']<=min_thresold)]

id                                               name    host_id  \
85        19601                perfect for a family or small group      74303   
496      174966              Luxury 2Bed/2.5Bath Central Park View     836168   
662      250801                 Heart & Soul of Greenwich Village     1314834   
762      273190          6 Bedroom Landmark West Village Townhouse     605463   
776      278145                 Large Room in a Huge NY apartment.    1452026   
...         ...                                                ...        ...   
48832  36450814                           FLATBUSH HANG OUT AND GO  267223765   
48839  36452721  Massage Spa. Stay overnight. Authors Artist dr...  274079964   
48852  36455809            Cozy Private Room in Bushwick, Brooklyn   74162901   
48867  36473044    The place you were dreaming for.(only for guys)  261338177   
48868  36473253                       Heaven for you(only for guy)  261338177   

          host_name neighbourhood_group      neighbourhood  latitude  \
85           Maggie            Brooklyn   Brooklyn Heights  40.69723   
496           Henry           Manhattan    Upper West Side  40.77350   
662           Rhona           Manhattan  Greenwich Village  40.73129   
762    West Village           Manhattan       West Village  40.73301   
776           Heidi              Queens            Astoria  40.77117   
...             ...                 ...                ...       ...   
48832        Jarmel            Brooklyn           Flatbush  40.64922   
48839       Richard            Brooklyn     Sheepshead Bay  40.59866   
48852     Christine            Brooklyn           Bushwick  40.69805   
48867         Diana            Brooklyn          Gravesend  40.59080   
48868         Diana            Brooklyn          Gravesend  40.59118   

       longitude        room_type  price  minimum_nights  number_of_reviews  \
85     -73.99268  Entire home/apt    800               1                 25   
496    -73.98697  Entire home/apt   2000              30                 30   
662    -73.99944  Entire home/apt    850               3                107   
762    -74.00268  Entire home/apt   1300               5                 28   
776    -73.91905     Private room     30               5                  3   
...          ...              ...    ...             ...                ...   
48832  -73.96078      Shared room     20               1                  0   
48839  -73.95661     Private room    800               1                  0   
48852  -73.92801     Private room     30               1                  1   
48867  -73.97116      Shared room     25               1                  0   
48868  -73.97119      Shared room     25               7                  0   

      last_review  reviews_per_month  calculated_host_listings_count  \
85     2016-08-04               0.24                               1   
496    2018-05-05               0.33                              11   
662    2019-05-23               1.15                               1   
762    2018-09-25               0.31                               4   
776    2017-06-20               0.03                               1   
...           ...                ...                             ...   
48832         NaN                NaN                               3   
48839         NaN                NaN                               1   
48852  2019-07-08               1.00                               1   
48867         NaN                NaN                               6   
48868         NaN                NaN                               6   

       availability_365  
85                    7  
496                   0  
662                 249  
762                 297  
776                   0  
...                 ...  
48832               363  
48839                23  
48852                 1  
48867               338  
48868               365  

[1151 rows x 16 columns]

In [13]:
# no outliers

df_no_outliers = df[(df['price']<max_thresold) & (df['price']>min_thresold)]

df_no_outliers.shape

(47744, 16)

We removed 1151 rows. While some may not be outliers, it's normally better to train a model without data on the far ends.